algorithms/ar-lreg/cpp/linear_regression.cpp

```diff
@@ -0,0 +1,184 @@
+/**
+ * One-dimensional Linear Regression implementation in C++.
+ * The generalized formula is target = weight * Variance(data) + bias.
+ * Above formula helps us to find the regression line by finding coefficent
+ * We can then find the target value to make a prediction based on given value of data.
+ */
+#include<iostream>
+#include<vector>
+
+using namespace std;
+/**
+ *  Generic class to contain one-dimensional
+ *  continous data and the target value.
+ */
+template<class T, class U>
+class DataPoint{
+public:
+// constructors
+DataPoint(T feature,U target);
+explicit DataPoint(T feature);
+// accessors and mutators
+void setFeature(const T feature);
+void setTarget(const U target);
+T getFeature() const;
+U getTarget() const;
+// destructors
+~DataPoint();
+private:
+  T feature;
+  U target;
+};
+// DataPoint Interface implementation
+template<class T, class U>
+DataPoint<T,U>::DataPoint(T feature, U target): feature(feature), target(target){}
+
+template<class T, class U>
+DataPoint<T,U>::DataPoint(T feature): feature(feature){}
+
+template<class T, class U>
+DataPoint<T,U>::~DataPoint(){}
+
+template<class T, class U>
+void DataPoint<T,U>::setFeature(const T feature){
+    this->feature = feature;
+}
+
+template<class T, class U>
+T DataPoint<T,U>::getFeature() const {
+    return feature;
+}
+
+template<class T, class U>
+void DataPoint<T,U>::setTarget(const U target){
+    this->target = target;
+}
+
+template<class T, class U>
+U DataPoint<T,U>::getTarget() const{
+    return this->target;
+}
+
+/**
+ * Linear Regression Interface.
+ * This interface takes DataPoint object in association. 
+ */
+template<typename T, typename U>
+class LinearRegression{
+public:
+   // Constructor
+  explicit LinearRegression(const vector<DataPoint<T,U>> &);
+   // Predicting the continous target value for the test-point.
+   DataPoint<T,U> fit(DataPoint<T,U> testPoint);
+   // Accessors
+   double inline getWeight() const;
+   double inline getBias() const;
+   // Destructors
+   ~LinearRegression();
+private:
+  LinearRegression();
+  vector<DataPoint<T,U>> dataset;
+  enum MEAN_TYPE{FEATURE=1,TARGET=2};
+  double weight;
+  double bias;
+  // Predicate helpers to calculate weight and bias
+  double calculateMean(const MEAN_TYPE) const; 
+  double calculateWeight() const;
+  double calculateBias() const;
+  // Mutators
+  void setWeight(const double weight);
+  void setBias(const double bias);
+};
+/**
+ *  LinearRegression Interface implementation.
+ */
+
+template<typename T, typename U>
+LinearRegression<T,U>::LinearRegression():weight(0.0),bias(0.0){}
+
+template<typename T, typename U>
+LinearRegression<T,U>::LinearRegression(const vector<DataPoint<T, U>> &data): dataset(data){}
```

Comment:
Codacy found an issue: [Member variable 'LinearRegression < int , int >::weight' is not initialized in the constructor.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6399935)